File: src/main/java/com/github/housepower/jdbc/data/type/DataTypeString.java
package com.github.housepower.jdbc.data.type;

import com.github.housepower.jdbc.data.IDataType;
import com.github.housepower.jdbc.misc.SQLLexer;
import com.github.housepower.jdbc.serializer.BinaryDeserializer;
import com.github.housepower.jdbc.serializer.BinarySerializer;

import java.io.IOException;
import java.sql.SQLException;
import java.sql.Types;

public class DataTypeString implements IDataType {

    @Override
    public String name() {
        return "String";
    }

    @Override
    public int sqlTypeId() {
        return Types.VARCHAR;
    }

    @Override
    public Object defaultValue() {
        return "";
    }

    @Override
    public Class javaTypeClass() {
        return String.class;
    }

    @Override
    public boolean nullable() {
        return false;
    }

    @Override
    public void serializeBinary(Object data, BinarySerializer serializer) throws SQLException, IOException {
        serializer.writeStringBinary((String) data);
    }

    @Override
    public Object deserializeBinary(BinaryDeserializer deserializer) throws SQLException, IOException {
        return deserializer.readStringBinary();
    }

    @Override
    public Object deserializeBinaryBulk(int rows, BinaryDeserializer deserializer)
        throws SQLException, IOException {
        String[] data = new String[rows];
        for (int row = 0; row < rows; row++) {
            data[row] = deserializer.readStringBinary();
        }
        return data;
    }

    @Override
    public Object deserializeTextQuoted(SQLLexer lexer) throws SQLException {
        return String.valueOf(lexer.stringLiteral());
    }
}

File: src/main/java/com/github/housepower/jdbc/data/type/DataTypeString.kt
package com.github.housepower.jdbc.data.type

import com.github.housepower.jdbc.data.IDataType
import com.github.housepower.jdbc.misc.SQLLexer
import com.github.housepower.jdbc.serializer.BinaryDeserializer
import com.github.housepower.jdbc.serializer.BinarySerializer
import java.io.IOException
import java.sql.SQLException
import java.sql.Types

class DataTypeString : IDataType {
    override fun name(): String {
        return "String"
    }

    override fun sqlTypeId(): Int {
        return Types.VARCHAR
    }

    override fun defaultValue(): Any {
        return ""
    }

    override fun javaTypeClass(): Class<*> {
        return String::class.java
    }

    override fun nullable(): Boolean {
        return false
    }


    override suspend fun serializeBinary(data: Any, serializer: BinarySerializer) {
        serializer.writeStringBinary((data as String))
    }


    override suspend fun deserializeBinary(deserializer: BinaryDeserializer): Any {
        return deserializer.readStringBinary()
    }

    
    override suspend fun deserializeBinaryBulk(rows: Int, deserializer: BinaryDeserializer): Any {
        val data = arrayOfNulls<String>(rows)
        for (row in 0 until rows) {
            data[row] = deserializer.readStringBinary()
        }
        return data
    }

    
    override fun deserializeTextQuoted(lexer: SQLLexer): Any {
        return lexer.stringLiteral().toString()
    }
}